upgrade.go

```diff
@@ -0,0 +1,500 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providertest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	jsonpb "google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	testutils "github.com/pulumi/pulumi-terraform-bridge/testing/x"
+	"github.com/pulumi/pulumi/pkg/v3/testing/integration"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+)
+
+type UpgradeTestMode int
+
+const (
+	UpgradeTestMode_Quick UpgradeTestMode = iota
+	UpgradeTestMode_PreviewOnly
+)
+
+func (m UpgradeTestMode) String() string {
+	switch m {
+	case UpgradeTestMode_PreviewOnly:
+		return "PreviewOnly"
+	case UpgradeTestMode_Quick:
+		return "Quick"
+	}
+	return "<unknown>"
+}
+
+func VerifyUpgrade(t *testing.T) *providerUpgradeBuilder {
+	return &providerUpgradeBuilder{
+		tt: t,
+		modes: map[UpgradeTestMode]string{
+			UpgradeTestMode_Quick:       "",
+			UpgradeTestMode_PreviewOnly: "",
+		},
+	}
+}
+
+type providerUpgradeBuilder struct {
+	tt                     *testing.T
+	resourceProviderServer pulumirpc.ResourceProviderServer
+	name                   string
+	program                string
+	modes                  map[UpgradeTestMode]string // skip reason by mode
+}
+
+func (b *providerUpgradeBuilder) Skip(
+	m UpgradeTestMode,
+	reason string,
+) *providerUpgradeBuilder {
+	require.NotEmpty(b.tt, reason, "reason cannot be empty")
+	b.modes[m] = reason
+	return b
+}
+
+func (b *providerUpgradeBuilder) WithProgram(dir string) *providerUpgradeBuilder {
```

Comment:
The builder pattern is a nod stylistically to @danielrbradley preferences. The Go way would be structs or functional options. I don't mind too much so long as we're consistent.